### PR TITLE
Fix #2804: Add ipv4 and ipv6 static addressing

### LINKED ISSUE #2804: Define static IP address for container
It would be useful to be able to define a static IP address for a container within the `docker-compose.yml`.

I see that `docker run --ip xxx.xxx.xxx.xxx` is now available in [docker 1.10.0-rc2](https://github.com/docker/docker/releases/tag/v1.10.0-rc2). More details in the merged pull request: https://github.com/docker/docker/pull/19001

Example:

    docker network create --driver=bridge --subnet=192.168.0.0/24 --gateway=192.168.0.1 mynet
    docker run --ip=192.168.0.10 --net=mynet -ti some_image

### PR DESCRIPTION
**Pull request by mdaue**
Fixes #2804

Signed-off-by: Matt Daue <mattdaue@gmail.com>

### CODE REVIEW

File: compose/service.py [hunk 1]
```diff
@@ -451,16 +452,29 @@ def start_container(self, container):
...
             if network in connected_networks:
                 self.client.disconnect_container_from_network(
                     container.id, network)
 
-            self.client.connect_container_to_network(
-                container.id, network,
-                aliases=list(self._get_aliases(container).union(aliases)),
-                links=self._get_links(False),
-            )
+            if docker.version_info >= (1, 8, 0):
```
Comment (dnephin) [inline compose/service.py]: We're going to set docker-py to a minimum version of 1.8.0, so we don't need this if/else branch.

For now you can set it to a git sha that contains the necessary commit in `requirements.txt`.

**Comment by dnephin** [comment]
Thanks for the PR!

Just the one comment about the docker-py version check, otherwise I think this looks like it's on the right track.

Still needs a test case and docs.

**Comment by clkao** [comment]
What happens when a service with static address gets scaled?  It should either be prevented or we should have some clever way for defining ip address corresponding to the service container index.

**Comment by mdaue** [comment]
Interesting problem, I think it should probably be prevented for now. The main use case for static IP addressing seems to be deterministic kernel filtering to operate services (and possibly DNS). If we circumvent these expected rulesets with automagic IP addressing, it may defeat the purpose.

We could look into doing something with IPVS and perhaps a round-robin scheduling approach of exposing the static IP and scaling using internal IPs via load balancing.

**Comment by mdaue** [comment]
tests pass locally with newer docker engine, I think janky fails with older docker engine: if \'client is newer than server\' in e.explanation:\nTypeError: \'str\' does not support the buffer interface\n'

File: .cache/v/cache/lastfailed [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (dnephin) [inline .cache/v/cache/lastfailed]: I believe this file was added by mistake. Please `git rm` it.

**Comment by dnephin** [comment]
This is looking good. Just one file that shouldn't be committed, and I think it would be good to squash the commits down to one or two commits.

**Comment by dnephin** [comment]
The test failure is unrelated.

**Comment by mdaue** [comment]
Removed inadvertent commit and squashed. Thanks!

**Comment by dnephin** [comment]
LGTM

File: requirements.txt [hunk 3]
```diff
@@ -3,7 +3,7 @@ cached-property==1.2.0
 dockerpty==0.4.1
 docopt==0.6.1
 enum34==1.0.4
-git+https://github.com/docker/docker-py.git@81d8caaf36159bf1accd86eab2e157bf8dd071a9#egg=docker-py
+git+https://github.com/mdaue/docker-py.git@d4a5bc4a86e9ab69368c80027439bca3ecc1e852#egg=docker-py
```
Comment (shin-) [inline requirements.txt]: Can you comment on that change? What is currently missing from docker-py? Is there a PR for it?

Comment (dnephin) [inline requirements.txt]: Oops, I missed this, good catch @shin- 

It depends on https://github.com/docker/docker-py/pull/936. Looks like that needs a rebase

Comment (shin-) [inline requirements.txt]: Ah, cool. I just merged it. We can use SHA `d8be3e0fce60fbe25be088b64bccbcee83effdb1` for now.

**Comment by mdaue** [comment]
Set requirements.txt to: git+https://github.com/docker/docker-py.git@d8be3e0fce60fbe25be088b64bccbcee83effdb1#egg=docker-py; this on @shin- suggestion

**Comment by shin-** [comment]
LGTM once green :+1:

**Comment by shin-** [comment]
It'll be in 1.7.0 - Target date simultaneous to the [Docker Engine 1.11.0 release](https://github.com/docker/docker/wiki#roadmap)

**Comment by CRogers** [comment]
Is there a beta version that we can use? 3 months is a long time to wait :(

**Comment by CRogers** [comment]
Oh, 04/07/2016 is american dates, only a couple of weeks to wait.

**Comment by mavenugo** [comment]
@shin- @dnephin we should prevent `scale` for the service with static ip-address. I couldn't find that change in this PR. Is that handled ?

**Comment by aanand** [comment]
@mavenugo from my manual testing, it looks like Engine will simply raise an error, which is fine. We could perhaps add a warning in Compose too though, to make the source of the problem clearer.

**Comment by loomchild** [comment]
First of all thanks a lot for implementing this feature.

While testing it I found an issue - when I try to start an application after stopping it, I get the following error:

    docker.errors.APIError: 409 Client Error: Conflict ("b'cannot create network br-777330a67cde (877e0da144dacec77255ffaeee34e99d4d5529179a4e11d1f8e2fb7ca72ad793): conflicts with network 777330a67cdee4012567000bfedcb8d4d77dbf9101f1667cbf4f3dc976895414 (br-777330a67cde): networks have overlapping IPv4'")

Apparently the new network cannot be created second time because it was already created during the first run. However, this works for networks with non-static IPs - new ones are simply ignored.

**Comment by mdaue** [comment]
@loomchild trying to duplicate your issue, could you please note your docker engine and docker-py versions?

**Comment by mdaue** [comment]
also, if you can share your docker-compose.yml that would be helpful - thanks!

**Comment by loomchild** [comment]
@mdaue I think I was too fast reporting a bug, I can't reproduce the error anymore. 

I was playing with different project names and it only occurs if I have two projects with different names trying to create network with the same range, which is expected I suppose.

As for the versions:

    Docker version 1.10.2, build c3959
    Docker-Compose commit 94b1862579049585c48a9f4039f275c02ebeb234 
    Docker-Py commit 7befe694bd21e3c54bb1d7825270ea4bd6864c13

Sorry for the noise.

**Comment by mdaue** [comment]
No problem! Two projects accessing the same network is something that would have to be addressed in the ipam-handling code portion I think.

**Comment by fordodone** [comment]
Thank you so much! works great engine 1.10.3 20f81dd compose 1.7.0dev 85e2fb6

**Comment by daniel-smith-kaazing** [comment]
Is there a new ETA? I see Docker 1.10.0 is out as of yesterday and there is a 1.7.0 RC for Compose that supports this feature. When do we expect the RC to become an R, so to speak?

**Comment by blah238** [comment]
@daniel-smith-kaazing https://github.com/docker/compose/releases/tag/1.7.0 was released today.

**Comment by daniel-smith-kaazing** [comment]
Great news! Thank you!

*Dan Smith*

Senior Support Engineer  |  Phone: 650.960.8155 | Mobile: 408.835.1167
daniel.smith@kaazing.com <dorris.obrien@kaazing.com>

*------------------------------------------------------------*

KAAZING >|<
When Real-time Matters
Kaazing.com <http://www.kaazing.com/>  |  Download
<http://kaazing.com/download.html>  |  Blog <http://blog.kaazing.com/>  |
Twitter <https://twitter.com/kaazing>
*http://kaazing.com <http://kaazing.com/about-us/careers/>*

**Pull request merged**